File: matrices/spiral-matrix-ii/spiral-matrix-ii.java
class Solution {
  public int[][] generateMatrix(int n) {
    int[][] grid = new int[n][n];

    int i = 0;
    int j = n - 1;
    int a = 0;
    int b = n - 1;

    int counter = 1;

    while (i <= j && a <= b) {
      for (int col = i; col <= j; col++) {
        grid[a][col] = counter++;
      }

      for (int row = a + 1; row <= b; row++) {
        grid[row][j] = counter++;
      }

      if (i < j || a < b) {
        for (int col = j - 1; col >= i + 1; col--) {
          grid[b][col] = counter++;
        }

        for (int row = b; row > a; row--) {
          grid[row][i] = counter++;
        }
      }

      i++;
      j--;
      a++;
      b--;
    }
    return grid;
  }
}

File: matrices/spiral-matrix-ii/spiral-matrix-ii.go
package spiralmatrixii

func generateMatrix(n int) [][]int {
	grid := make([][]int, n)

	for i := range grid {
		grid[i] = make([]int, n)
	}

	i := 0
	j := n - 1
	a := 0
	b := n - 1

	counter := 1

	for i <= j && a <= b {
		for col := i; col <= j; col++ {
			grid[a][col] = counter
			counter++
		}

		for row := a + 1; row <= b; row++ {
			grid[row][j] = counter
			counter++
		}

		if i < j || a < b {
			for col := j - 1; col >= i+1; col-- {
				grid[b][col] = counter
				counter++
			}

			for row := b; row > a; row-- {
				grid[row][i] = counter
				counter++
			}
		}
		i++
		j--
		a++
		b--
	}
	return grid
}
